tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -192,6 +193,85 @@ void ROOT::Experimental::Detail::RPageSinkFile::ReleasePage(RPage &page)
 ////////////////////////////////////////////////////////////////////////////////
 
 
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+class RPageCache {
+private:
+   // Doubly linked-list for random-access removal
+   std::list<RPage> fPages;
+
+   /// Clear the page's column id and element size but reuse the sized buffer.
+   RPage RecyclePage(const RPage &page) const {
+      return RPage(
+         kInvalidColumnId, page.GetBuffer(), page.GetCapacity(), 0 /* elementSize */
+      );
+   }
+
+public:
+   RPageCache() = default;
+   RPageCache(const RPageCache&) = delete;
+   RPageCache& operator=(const RPageCache&) = delete;
+   RPageCache(RPageCache&&) = default;
+   RPageCache& operator=(RPageCache&&) = default;
+
+   /// Deallocate the page's memory buffer.
+   void DeletePage(const RPage &page) {
+      delete[] reinterpret_cast<unsigned char *>(page.GetBuffer());
+   }
+
+   /// Return a page to the cache for reuse. Page memory ownership is passed to the cache.
+   /// If the cache is full, the smallest memory buffer is deallocated.
+   void ReturnPage(const RPage &page) {
+      fPages.push_back(RecyclePage(page));
+      if (fPages.size() <= 32) {
+         return;
+      }
+      // If the cache is getting too large, deallocate the smallest page
+      auto min_elt = std::min_element(fPages.begin(), fPages.end(),
+         [](const auto &a, const auto &b) { return a.GetCapacity() < b.GetCapacity(); });
+      DeletePage(*min_elt);
+      fPages.erase(min_elt);
+   }
+
+   /// Request a page from the cache with at least the specified `capacity`. Ownership of the
+   /// page memory is passed to the caller. If there isn't a suitable page in the cache, a
+   /// new allocation is made. The caller is responsible for properly setting the page's
+   /// column id and element size.
+   RPage RequestPage(std::size_t capacity) {
+      auto it = std::find_if(fPages.begin(), fPages.end(), [&](const auto &p) {
+         return p.GetCapacity() >= capacity;
+      });
```

Comment:
As said above, we might also benefit here of using a slightly different approach, e.g. ordered list, or a bucket-based data structure.